compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -62,6 +62,143 @@ pub(super) fn predicates_of(tcx: TyCtxt<'_>, def_id: DefId) -> ty::GenericPredic
 fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::GenericPredicates<'_> {
     use rustc_hir::*;
 
+    match tcx.opt_rpitit_info(def_id.to_def_id()) {
+        Some(ImplTraitInTraitData::Trait { fn_def_id, opaque_def_id }) => {
+            let mut predicates = Vec::new();
+            // FIXME we need a query and we need bidir_outlives metadata
+            if opaque_def_id.is_local() {
+                let icx = ItemCtxt::new(tcx, def_id);
+                let hir_id = tcx.hir().local_def_id_to_hir_id(opaque_def_id.expect_local());
+                let node = tcx.hir().get(hir_id);
+                let Node::Item(&Item { kind: ItemKind::OpaqueTy(OpaqueTy { bidir_outlives, .. }), .. }) = node else {
+                    bug!("unexpected {node:?}")
+                };
+
+                for (arg, dup_def) in bidir_outlives {
+                    let orig_region = icx.astconv().ast_region_to_region(arg, None);
+                    if !matches!(orig_region.kind(), ty::ReEarlyBound(..)) {
+                        // Only early-bound regions can point to the original generic parameter.
+                        continue;
+                    }
+
+                    let generics = tcx.generics_of(def_id);
+
+                    let Some(dup_index) = generics.param_def_id_to_index(tcx, dup_def.to_def_id()) else { bug!() };
+
+                    let dup_region = ty::Region::new_early_bound(
+                        tcx,
+                        ty::EarlyBoundRegion {
+                            def_id: dup_def.to_def_id(),
+                            index: dup_index,
+                            name: arg.ident.name,
+                        },
+                    );
+
+                    predicates.push((
+                        ty::Binder::dummy(ty::PredicateKind::Clause(ty::Clause::RegionOutlives(
+                            ty::OutlivesPredicate(orig_region, dup_region),
+                        )))
+                        .to_predicate(icx.tcx),
+                        arg.ident.span,
+                    ));
+                    predicates.push((
+                        ty::Binder::dummy(ty::PredicateKind::Clause(ty::Clause::RegionOutlives(
+                            ty::OutlivesPredicate(dup_region, orig_region),
+                        )))
+                        .to_predicate(icx.tcx),
+                        arg.ident.span,
+                    ));
+                }
+                debug!(?predicates);
+            }
+
+            let trait_def_id = tcx.parent(fn_def_id);
+            return ty::GenericPredicates {
+                parent: Some(trait_def_id),
+                predicates: tcx.arena.alloc_from_iter(predicates),
+            };
+        }
+
+        Some(ImplTraitInTraitData::Impl { fn_def_id }) => {
```

Comment:
I think you can actually just take the predicates from the trait GAT and substitute it with `ty::InternalSubsts::identity_for_item(assoc_item).rebase_onto(impl_def_id, tcx.impl_trait_ref(impl_def_id).unwrap())` instead of doing anything else here.

Reply:
Ohh, I've been scratching my head today and found the solution for this problem only to see that you've shared it yesterday. lol, anyway it was a good exercise.